### ISSUE TITLE
default_collate transposing orientation of list

### ISSUE THREAD

**Issue by piercefreeman** [pytorch/pytorch#23141]
## 🐛 Bug / Unexpected Behavior

I have a custom subclass of `Dataset` where each datapoint returns a dictionary with a nested list within it.  Say the minimum sequence length is N over a particular batch of size M.  When I batch these datapoints together using `DataLoader`, it returns the batched sequence as NxM.

According to the documentation for `default_collate`, it should orient the batch of these datapoints with the "outer dimension batch size".  Here that seems to be violated.

## To Reproduce

```
from torch.utils.data import Dataset
from torch.utils.data import DataLoader
import numpy as np

class CustomDataset(Dataset):
    def __len__(self):
        return 2

    def __getitem__(self, idx):
        return {
            "foo": np.array([1, 2, 3]),
            "bar": ["X"] * (idx+1),
        }

training = CustomDataset()

for batch in DataLoader(training, batch_size=2):
    print(batch)
```

Here we see the result as:

```
{
  'foo': tensor(
    [
      [1, 2, 3],
      [1, 2, 3]
    ]
  ),
  'bar': [
      ('X', 'X'),
    ]
}
```

## Expected behavior

I would expect the output to be:

```
{
  'foo': tensor(
    [
      [1, 2, 3],
      [1, 2, 3]
    ]
  ),
  'bar': [
      ('X'),
      ('X', 'X'),
    ]
}
```

## Environment

PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.14.3
GCC version: Could not collect
CMake version: version 3.14.3

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.16.4
[pip3] pytorch-pretrained-bert==0.6.2
[pip3] torch==1.1.0
[conda] Could not collect

**Comment by SsnL** [comment]
This is a bit of a weird case... 

When `default_collate` tries to collate a batch like `[('X',), ('X', 'X')]`, it collates each element of the list, and since the second element isn't present in all samples, that is discarded. 

This certainly could be unexpected, but I'm not sure what the solution should be though. What you suggest is treating the list itself as the data sampler, rather than a collection as the data sample, which is inconsistent with the documentation and intended use of `default_collate`. It also doesn't make sense to just arbitrarily put a placeholder (like `None`) and return `[('X', ’X'), (None, 'X')]` either. 

I would say that `default_collate` should error out in this case. What do you think?

**Comment by piercefreeman** [comment]
@SsnL Your argument against automatic `None` padding makes sense to me as well.  I think we should throw an error if not of the same length.

In my mind, the bigger issue here is that the outer dimension of collections isn't the batch size, but instead is the minimum sampled length (`NxM`).  To me that is unexpected behavior and should align with the behavior of batching tensors, where we have `MxN` dimensionality.

**Comment by soumith** [comment]
yea erroring out here makes sense.

**Comment by choidongyeon** [comment]
Hi, I would like to work on this issue. Can anyone help me get started?

**Comment by choidongyeon** [comment]
@SsnL, I think this can be closed now.
